plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/action/DefaultActionDispatcher.java

```diff
@@ -15,44 +15,198 @@
 import org.eclipse.glsp.api.registry.ActionHandlerRegistry;
 
 import com.google.inject.Inject;
+import com.google.inject.Provider;
 
-public class DefaultActionDispatcher implements ActionDispatcher {
-   private static Logger LOG = Logger.getLogger(DefaultActionDispatcher.class);
+/**
+ * <p>
+ * An ActionDispatcher that executes all handlers in the same thread. The dispatcher's
+ * public methods can be invoked from any thread, e.g. from background jobs running on
+ * the server.
+ * </p>
+ */
+public class DefaultActionDispatcher implements ActionDispatcher, ClientSessionListener {
 
+   private static final Logger LOG = Logger.getLogger(DefaultActionDispatcher.class);
+
+   private static final AtomicInteger COUNT = new AtomicInteger(0);
+
+   @Inject
+   private ActionHandlerRegistry actionHandlerRegistry;
+
+   @Inject
+   private ModelStateProvider modelStateProvider;
+
+   private final ClientSessionManager clientSessionManager;
+
+   private final String name;
+
+   private final Thread thread;
+
+   private final BlockingQueue<ActionMessage> actionsQueue = new ArrayBlockingQueue<>(100, true);
+
+   // Results will be placed in the map when the action dispatcher receives a new message (From arbitrary threads),
+   // and will be removed from the dispatcher's thread.
+   private final Map<ActionMessage, CompletableFuture<Void>> results = Collections.synchronizedMap(new HashMap<>());
+
+   // Use a provider, as the GLSPClient is probably not created yet. We won't receive
+   // any message until it's ready anyway.
    @Inject
-   protected ActionHandlerRegistry actionHandlerRegistry;
+   private Provider<GLSPClient> client;
 
    @Inject
-   protected ModelStateProvider modelStateProvider;
+   public DefaultActionDispatcher(final ClientSessionManager clientSessionManager) {
+      this.name = getClass().getSimpleName() + " " + COUNT.incrementAndGet();
+      this.clientSessionManager = clientSessionManager;
+      this.clientSessionManager.addListener(this);
+      this.thread = new Thread(this::runThread);
+      this.thread.setName(this.name);
+      this.thread.setDaemon(true);
+      this.thread.start();
+   }
 
    @Override
-   public void dispatch(final String clientId, final Action action) {
-      List<ActionHandler> actionHandlers = actionHandlerRegistry.get(action);
-      if (actionHandlers.isEmpty()) {
-         LOG.warn("No handler registered for action: " + action);
+   public CompletableFuture<Void> dispatch(final String clientId, final Action action) {
+      return dispatch(new ActionMessage(clientId, action));
+   }
+
+   @Override
+   public CompletableFuture<Void> dispatch(final ActionMessage message) {
+      if (message == null) {
+         String errorMsg = String.format("Received a null message in DefaultActionDispatcher: %s", name);
+         throw new IllegalArgumentException(errorMsg);
+      }
+      final CompletableFuture<Void> result = results.putIfAbsent(message, new CompletableFuture<Void>());
```

Comment:
This causes  NullPointerExceptions. Result is null,if there was no previous mapping for the given key and is returned instead of valid CompletableFuture<void>. Maybe use computeIfAbsent instead